pkg/helm/charts.go

```diff
@@ -0,0 +1,201 @@
+		}
+	}
+
+	// return Charts handle
+	charts := &Charts{
+		Manifest:    c,
+		projectRoot: abs,
+
+		// default to ExecHelm, but allow injecting from the outside
+		Helm: ExecHelm{},
+	}
+	return charts, nil
+}
+
+// Charts exposes the central Chartfile management functions
+type Charts struct {
+	// Manifest are the chartfile.yaml contents. It holds data about the developers intentions
+	Manifest Chartfile
+
+	// projectRoot is the enclosing directory of chartfile.yaml
+	projectRoot string
+
+	// Helm is the helm implementation underneath. ExecHelm is the default, but
+	// any implementation of the Helm interface may be used
+	Helm Helm
+}
+
+// ChartDir returns the directory pulled charts are saved in
+func (c Charts) ChartDir() string {
+	return filepath.Join(c.projectRoot, c.Manifest.Directory)
+}
+
+// ManifestFile returns the full path to the chartfile.yaml
+func (c Charts) ManifestFile() string {
+	return filepath.Join(c.projectRoot, Filename)
+}
+
+// Vendor pulls all Charts specified in the manifest into the local charts
+// directory. It fetches the repository index before doing so.
+func (c Charts) Vendor() error {
+	dir := c.ChartDir()
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return err
+	}
+
+	log.Println("Syncing Repositories ...")
+	if err := c.Helm.RepoUpdate(Opts{Repositories: c.Manifest.Repositories}); err != nil {
+		return err
+	}
+
+	log.Println("Pulling Charts ...")
+	for _, r := range c.Manifest.Requires {
+		err := c.Helm.Pull(r.Chart, r.Version.String(), PullOpts{
+			Destination: dir,
+			Opts:        Opts{Repositories: c.Manifest.Repositories},
+		})
+		if err != nil {
+			return err
+		}
+
+		log.Printf(" %s@%s", r.Chart, r.Version.String())
+	}
+
+	return nil
+}
+
+// Add adds every Chart in reqs to the Manifest after validation, and runs
+// Vendor afterwards
+func (c Charts) Add(reqs []string) error {
+	log.Printf("Adding %v Charts ...", len(reqs))
+
+	skip := func(s string, err error) {
+		log.Printf(" Skipping %s: %s.", s, err)
+	}
+
+	// parse new charts, append in memory
+	added := 0
+	for _, s := range reqs {
+		r, err := parseReq(s)
+		if err != nil {
+			skip(s, err)
+			continue
+		}
+
+		if c.Manifest.Requires.Has(*r) {
+			skip(s, fmt.Errorf("already exists"))
+			continue
+		}
+
+		c.Manifest.Requires = append(c.Manifest.Requires, *r)
+		added++
+		log.Println(" OK:", s)
+	}
+
+	// write out
+	if err := write(c.Manifest, c.ManifestFile()); err != nil {
+		return err
+	}
+
+	// skipped some? fail then
+	if added != len(reqs) {
+		return fmt.Errorf("%v Charts were skipped. Please check above logs for details", len(reqs)-added)
+	}
+
+	// worked fine? vendor it
+	log.Printf("Added %v Charts to helmfile.yaml. Vendoring ...", added)
+	return c.Vendor()
+}
+
+func InitChartfile(path string) (*Charts, error) {
+	c := Chartfile{
+		Version: Version,
+		Repositories: []Repo{{
+			Name: "stable",
+			URL:  "https://kubernetes-charts.storage.googleapis.com",
```

Comment:
do we want this? Being deprecated soon?

Reply:
Currently it's the primary source of truth. Soon is not yet deprecated :)

We will remove as soon as deprecated